edb/server/protocol/ext/auth.py

```diff
@@ -0,0 +1,282 @@
+        state=state_token.serialize(), redirect_uri=redirect_uri
+    )
+
+    response.status = http.HTTPStatus.FOUND
+    response.custom_headers["Location"] = auth_url
+    response.close_connection = True
+
+
+async def handle_auth_callback(request, response):
+    return
+
+
+async def handle_request(request, response, db, args):
+    try:
+        # Set up routing to the appropriate handler
+        if args[0] == "authorize":
+            provider_name: str | None = _get_search_param(
+                request.url.query.decode("ascii"), "provider"
+            )
+            if provider_name is None:
+                raise errors.InternalServerError("No provider specified")
+
+            redirect_uri = f"{_get_extension_path(request.url)}/callback"
+
+            return redirect_to_auth_provider(
+                response=response,
+                redirect_uri=redirect_uri,
+                iss=request.host.decode(),
+                provider_name=provider_name,
+                db_config=db.db_config,
+            )
+        elif args[0] == "callback":
+            await handle_auth_callback(request, response)
+        else:
+            raise errors.BackendError
+    except Exception as ex:
+        if debug.flags.server:
+            markup.dump(ex)
+
+        # XXX Fix this when LSP "location" objects are implemented
+        ex_type = errors.InternalServerError
+
+        _response_error(
+            response, http.HTTPStatus.INTERNAL_SERVER_ERROR, str(ex), ex_type
+        )
+
+
+def _response_error(response, status, message, ex_type):
+    err_dct = {
+        "message": message,
+        "type": str(ex_type.__name__),
+        "code": ex_type.get_code(),
+    }
+
+    response.body = json.dumps({"error": err_dct}).encode()
+    response.status = status
+    response.close_connection = True
+
+
+# OAuth 2 Provider for GitHub
+class GitHubProvider(BaseProvider):
+    def __init__(self, client_id, client_secret):
+        super().__init__("github", client_id, client_secret)
+
+    def get_code_url(
+        self, state: str, redirect_uri: str, scope: str = "read:user"
+    ) -> str:
+        params = {
+            "client_id": self.client_id,
+            "scope": scope,
+            "state": state,
+            "redirect_uri": redirect_uri,
+        }
+        encoded = urllib.parse.urlencode(params)
+        return f"https://github.com/login/oauth/authorize?{encoded}"
+
+    async def exchange_access_token(self, code, state):
+        # Check state value
+        # TODO: Look up state value from FlowState object
+        # flow_state = await db.get_flow_state(state, "github")
+        # if flow_state is None:
+        #    raise errors.UnauthorizedError("invalid state value")
+        data = {
+            'grant_type': 'authorization_code',
+            'code': code,
+            'state': state,
+            "redirect_uri": "http://localhost:8080/auth/callback",  # TODO: Get correct db-namespaced URL
```

Comment:
I think you can pass the `server` object from the `protocol.pyx` to get host/port. And yeah, let's DB-namespace the URL.

Reply:
Yeah, I ended up building this dynamically from the request URL in that `_get_extension_path` function, but as you pointed out later, it's pretty fragile. I'll move the extension's base route building to the router logic itself, it seems like that is really the right place that owns the route up to this point.

Reply:
⚡ [`6d508ea` (#5816)](https://github.com/edgedb/edgedb/pull/5816/commits/6d508eab75c7e433d5ad3b7137910a0e5ab63df7) @1st1 

How does that look?